File: 160-intersection-of-two-linked-lists/160-intersection-of-two-linked-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        auto tempA = headA,tempB=headB;
        while(tempA&&tempB){
            tempA = tempA->next;
            tempB = tempB->next;
        }
        ListNode* temp;
        if(!tempA){
            tempA = headB;
            while(tempB){
                tempB = tempB->next;
                tempA = tempA->next;
            }
            tempB = headA;
        }
        else{
            tempB = headA;
            while(tempA){
                tempB = tempB->next;
                tempA = tempA->next;
            }
            tempA = headB;
        }
        
        while(tempB&&tempA){
            if(tempB==tempA) return tempA;
            tempB = tempB->next;
            tempA = tempA->next;
        }
        
        return tempA;
        
    }
};

File: 160-intersection-of-two-linked-lists/160-intersection-of-two-linked-lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        
        t1,t2 = headA,headB
        
        while t1 and t2:
            t1 = t1.next
            t2 = t2.next
            
        if t1==None:
            t1 = headB
            backup = headA
        else:
            t2 = headA
            backup = headB
        
        while t1 and t2:
            t1 = t1.next
            t2 = t2.next
            
        if t1==None:
            t1 = backup
        else:
            t2 = backup
        
        while t1 and t2:
            if t1==t2: return t1
            t1 = t1.next
            t2 = t2.next
        
        return None
            